File: FateGrandAutomata/CutoutManager.cs
using Android.Content;
using Android.OS;
using Android.Util;
using Android.Views;
using CoreAutomata;
using Java.Interop;

namespace FateGrandAutomata
{
    public static class CutoutManager
    {
        static bool _cutoutFound;

        static (int L, int T, int R, int B)? _cutoutVal;

        public static void ApplyCutout(MainActivity MainActivity)
        {
            if (_cutoutFound)
                return;

            // Android P added support for display cutouts
            if (Build.VERSION.SdkInt < BuildVersionCodes.P)
                return;

            var cutout = MainActivity.Window.DecorView.RootWindowInsets.DisplayCutout;
            if (cutout == null)
                return;

            var l = cutout.SafeInsetLeft;
            var t = cutout.SafeInsetTop;
            var r = cutout.SafeInsetRight;
            var b = cutout.SafeInsetBottom;

            // Check if there is a cutout
            if (!(l == 0 && t == 0 && r == 0 && b == 0))
            {
                var wm = MainActivity.GetSystemService(Context.WindowService).JavaCast<IWindowManager>();
                var rotation = wm.DefaultDisplay.Rotation;

                // Store the cutout for Portrait orientation of device
                switch (rotation)
                {
                    case SurfaceOrientation.Rotation90:
                        (l, t, r, b) = (b, l, t, r);
                        break;

                    case SurfaceOrientation.Rotation180:
                        (l, t, r, b) = (r, b, l, t);
                        break;

                    case SurfaceOrientation.Rotation270:
                        (l, t, r, b) = (t, r, b, l);
                        break;
                }

                _cutoutVal = (l, t, r, b);
            }

            _cutoutFound = true;
        }

        static (int L, int T, int R, int B)? GetCutout(SurfaceOrientation Rotation)
        {
            if (!GameAreaManager.AutoGameArea)
                return null;

            var cutout = _cutoutVal;

            if (cutout == null)
                return null;

            var (l, t, r, b) = cutout.Value;

            // Consider current orientation of screen
            switch (Rotation)
            {
                case SurfaceOrientation.Rotation90:
                    (l, t, r, b) = (t, r, b, l);
                    break;

                case SurfaceOrientation.Rotation180:
                    (l, t, r, b) = (r, b, l, t);
                    break;

                case SurfaceOrientation.Rotation270:
                    (l, t, r, b) = (b, l, t, r);
                    break;
            }

            return (l, t, r, b);
        }

        public static Region GetCutoutAppliedRegion(Context Context)
        {
            var metrics = new DisplayMetrics();
            var wm = Context.GetSystemService(Context.WindowService).JavaCast<IWindowManager>();

            wm.DefaultDisplay.GetRealMetrics(metrics);

            var w = metrics.WidthPixels;
            var h = metrics.HeightPixels;
            var x = 0;
            var y = 0;

            var cutout = GetCutout(wm.DefaultDisplay.Rotation);

            if (cutout != null)
            {
                var (l, t, r, b) = cutout.Value;

                x = l;
                y = t;
                w -= l + r;
                h -= t + b;
            }

            return new Region(x, y, w, h);
        }
    }
}

File: app/src/main/java/com/mathewsachin/fategrandautomata/util/CutoutManager.kt
package com.mathewsachin.fategrandautomata.util

import android.app.Activity
import android.os.Build
import android.view.Surface
import com.mathewsachin.fategrandautomata.prefs.core.GameAreaMode
import com.mathewsachin.fategrandautomata.prefs.core.PrefsCore
import com.mathewsachin.libautomata.Region
import com.mathewsachin.libautomata.Size
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CutoutManager @Inject constructor(
    private val display: DisplayHelper,
    private val prefsCore: PrefsCore,
) {
    private data class Cutout(val L: Int = 0, val T: Int = 0, val R: Int = 0, val B: Int = 0) {
        companion object {
            val NoCutouts = Cutout()
        }
    }

    private var cutoutFound = false
    private var cutoutValue = Cutout.NoCutouts

    fun applyCutout(activity: Activity) {
        if (cutoutFound) {
            return
        }

        // Android P added support for display cutouts
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P) {
            cutoutFound = true
            return
        }

        val displayCutout = activity.window.decorView.rootWindowInsets.displayCutout
        if (displayCutout == null) {
            cutoutFound = true
            return
        }

        val cutout = Cutout(
            displayCutout.safeInsetLeft,
            displayCutout.safeInsetTop,
            displayCutout.safeInsetRight,
            displayCutout.safeInsetBottom
        )

        // Check if there is a cutout
        if (cutout != Cutout.NoCutouts) {
            val rotation = display.rotation

            // Store the cutout for Portrait orientation of device
            val (l, t, r, b) = cutout
            cutoutValue = when (rotation) {
                Surface.ROTATION_90 -> Cutout(b, l, t, r)
                Surface.ROTATION_180 -> Cutout(r, b, l, t)
                Surface.ROTATION_270 -> Cutout(t, r, b, l)
                else -> cutout
            }
        }

        cutoutFound = true
        Timber.d("Detected display cutout: $cutoutValue")
    }

    private fun getScreenSize(): Size {
        val metrics = display.metrics

        return Size(metrics.widthPixels, metrics.heightPixels)
    }

    fun getCutoutAppliedRegion(): Region {
        val (w, h) = getScreenSize()

        return when (prefsCore.gameAreaMode.get()) {
            GameAreaMode.Default -> Region(0, 0, w, h)
            GameAreaMode.Duo -> Region(192, 0, 2400, h)
            GameAreaMode.Custom -> {
                val l = prefsCore.gameOffsetLeft.get()
                val t = prefsCore.gameOffsetTop.get()
                val r = prefsCore.gameOffsetRight.get()
                val b = prefsCore.gameOffsetBottom.get()

                // if the camera is on the right, use the right offset as x
                val x = if (display.rotation == Surface.ROTATION_270) r else l
                Region(x, t, w - l - r, h - t - b)
            }
        }
    }
}
